libs/compute_cuda/include/hpx/compute/cuda/target.hpp

```diff
@@ -32,6 +35,149 @@
 #include <hpx/config/warnings_prefix.hpp>
 
 namespace hpx { namespace compute { namespace cuda {
+
+    ///////////////////////////////////////////////////////////////////////////
+    namespace detail {
+        struct runtime_registration_wrapper
+        {
+            runtime_registration_wrapper(hpx::runtime* rt)
+              : rt_(rt)
+            {
+                HPX_ASSERT(rt);
+
+                // Register this thread with HPX, this should be done once for
+                // each external OS-thread intended to invoke HPX functionality.
+                // Calling this function more than once on the same thread will
+                // report an error.
+                hpx::error_code ec(hpx::lightweight);    // ignore errors
+                hpx::register_thread(rt_, "cuda", ec);
+            }
+            ~runtime_registration_wrapper()
+            {
+                // Unregister the thread from HPX, this should be done once in
+                // the end before the external thread exists.
+                hpx::unregister_thread(rt_);
+            }
+
+            hpx::runtime* rt_;
+        };
+
+        template <typename Allocator>
+        struct future_data;
+
+        template <typename Allocator>
+        struct release_on_exit
+        {
+            release_on_exit(future_data<Allocator>* data)
+              : data_(data)
+            {
+            }
+
+            ~release_on_exit()
+            {
+                // release the shared state
+                lcos::detail::intrusive_ptr_release(data_);
+            }
+
+            future_data<Allocator>* data_;
+        };
+
+        template <typename Allocator>
+        struct future_data
+          : lcos::detail::future_data_allocator<void, Allocator>
+        {
+            using init_no_addref =
+                typename lcos::detail::future_data_allocator<void,
+                    Allocator>::init_no_addref;
+            using other_allocator = typename std::allocator_traits<
+                Allocator>::template rebind_alloc<future_data>;
+
+            static void CUDART_CB stream_callback(
+                cudaStream_t stream, cudaError_t error, void* user_data)
+            {
```

Comment:
This function could be HPX_EXPORT'ed and moved to the .cpp file

Reply:
Wait, can it really? It accesses the `rt_` member from `future_data` which means that it's effectively also templated on `Allocator`. What am I missing?

Reply:
It accesses the member through the pointer it receives. The whole function does not depend on the template argument and could be moved outside of the class.

Reply:
Right, it can be moved outside the class (it's static after all) but not into the cpp file. It still needs to know to what it should cast the `void*` so it would still need to be templated on `Allocator` (or on the full `future_data` type). Am I still missing something?

Reply:
Yes, you're certainly right. Sorry for the confusion.